runtime-light/stdlib/string/pcre2-functions.h

```diff
@@ -0,0 +1,268 @@
+
+namespace kphp::pcre2 {
+
+struct error {
+  int32_t code{};
+};
+
+struct compile_error : error {
+  size_t offset{};
+};
+
+struct group_name {
+  std::string_view name;
+  size_t index;
+};
+
+class group_name_iterator {
+public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = group_name;
+  using difference_type = std::ptrdiff_t;
+  using pointer = group_name*;
+  using reference = group_name;
+
+  group_name_iterator(const PCRE2_UCHAR8* current_entry, size_t entry_size) noexcept
+      : m_ptr{current_entry},
+        m_entry_size{entry_size} {}
+
+  group_name operator*() const noexcept {
+    const auto index{static_cast<size_t>(m_ptr[0] << 8 | m_ptr[1])};
+    const auto* name_ptr{reinterpret_cast<const char*>(std::next(m_ptr, 2))};
+    return {.name = std::string_view{name_ptr}, .index = index};
+  }
+
+  group_name_iterator& operator++() noexcept {
+    std::advance(m_ptr, m_entry_size);
+    return *this;
+  }
+
+  group_name_iterator operator++(int) noexcept {
+    group_name_iterator tmp{*this};
+    ++*this;
+    return tmp;
+  }
+
+  bool operator==(const group_name_iterator& other) const noexcept {
+    return m_ptr == other.m_ptr;
+  }
+
+private:
+  const PCRE2_UCHAR8* m_ptr;
+  size_t m_entry_size;
+};
+
+class regex {
+public:
+  friend class matcher;
+
+  static std::expected<regex, compile_error> compile(std::string_view pattern, uint32_t options = 0, pcre2_compile_context_8* ctx = nullptr) noexcept;
+
+  struct group_name_range {
+    group_name_iterator b;
+    group_name_iterator e;
+
+    group_name_iterator begin() const noexcept {
+      return b;
+    }
+    group_name_iterator end() const noexcept {
+      return e;
+    }
+
+    bool empty() const noexcept {
+      return b == e;
+    }
+  };
+
+  group_name_range names() const noexcept;
+
+  uint32_t capture_count() const noexcept {
+    uint32_t count{};
+    pcre2_pattern_info_8(m_code.get(), PCRE2_INFO_CAPTURECOUNT, &count);
+    return count;
+  }
+
+  uint32_t name_count() const noexcept {
+    uint32_t count{};
+    pcre2_pattern_info_8(m_code.get(), PCRE2_INFO_NAMECOUNT, &count);
+    return count;
+  }
+
+  bool is_utf() const noexcept {
+    uint32_t compile_options{};
+    pcre2_pattern_info_8(m_code.get(), PCRE2_INFO_ARGOPTIONS, &compile_options);
+    return (compile_options & PCRE2_UTF) != 0;
+  }
+
+  /**
+   * @param buffer_len Input: capacity of buffer. Output: actual length of result.
+   * @return expected<size_t, error>: The number of replacements (should be 1).
+   */
+  std::expected<size_t, error> substitute_match(std::string_view subject, pcre2_match_data_8& data, std::string_view replacement, char* buffer,
+                                                size_t& buffer_len, uint32_t match_options, pcre2_match_context_8* ctx = nullptr) const noexcept;
+
+private:
+  explicit regex(pcre2_code_8& code)
+      : m_code{std::addressof(code), pcre2_code_free_8} {}
+  regex_pcre2_code_t m_code;
```

Comment:
I wonder if it's better for readability to move those types into `kphp::pcre2` namespace. For example, it could be `kphp::pcre2::code`